### PR TITLE
Adding eq4 for StTriggerData

### PR DESCRIPTION
**Pull request by akioogawa**
There is new crate called EQ4 in the trigger system now. This updates StTriggerData to read EQ4 crate.

### CODE REVIEW

File: StRoot/StEvent/StTriggerData2022.cxx [hunk 1]
```diff
@@ -1403,16 +1407,17 @@ unsigned short StTriggerData2022::mtd4AtAddress(int address, int prepost) const
 unsigned short StTriggerData2022::fmsADC(int crt, int adr, int ch, int prepost) const
 {
     int buffer = prepostAddress(prepost);
-    if (buffer >= 0 && crt>=1 && crt<=7 && adr>=0 && adr<16 && ch>=0 && ch<=31){
+    if (buffer >= 0 && crt>=1 && crt<=8 && adr>=0 && adr<16 && ch>=0 && ch<=31){
         switch(crt){
 	    case 0: return bbq[buffer][adr][ch];
-	    case 1: return qt1[buffer][adr][ch];
```
Comment (klendathu2k) [inline StRoot/StEvent/StTriggerData2022.cxx:1409]: The "if" statement which precedes  the switch statement will allow any value for crt between 1 and 8... but...

1) Why is there a case 0 that cannot be reached?
2) Cases 1 to 4 are not handled.  Strongly recommend a default statement (ending in at least a stern warning, if not an outright assert) to catch errors elsewehere in the code.

File: StRoot/StEvent/StTriggerData2022.cxx [hunk 2]
```diff
@@ -1421,16 +1426,17 @@ unsigned short StTriggerData2022::fmsADC(int crt, int adr, int ch, int prepost)
 unsigned short StTriggerData2022::fmsTDC(int crt, int adr, int ch, int prepost) const
 {
     int buffer = prepostAddress(prepost);
-    if (buffer >= 0 && crt>=0 && crt<=7 && adr>=0 && adr<16 && ch>=0 && ch<=31){
+    if (buffer >= 0 && crt>=0 && crt<=8 && adr>=0 && adr<16 && ch>=0 && ch<=31){
         switch(crt){
             case 0: return tbbq[buffer][adr][ch];
-            case 1: return tqt1[buffer][adr][ch];
```
Comment (klendathu2k) [inline StRoot/StEvent/StTriggerData2022.cxx:1427]: Same comment (although this time case 0 can be reached).

**Review by klendathu2k — APPROVED** [review]
I approve.  One recommendation is made for consideration.

File: StRoot/StEvent/StTriggerData2022.cxx [hunk 3]
```diff
@@ -144,15 +144,17 @@ void StTriggerData2022::readData(const TriggerDataBlk2022* data, int bs) {
...
     memset(mEQ2,0,sizeof(mEQ2));
+    memset(mEQ4,0,sizeof(mEQ4));
     memset(mxq,0,sizeof(mxq)); memset(tmxq,0,sizeof(tmxq));
     memset(eq3,0,sizeof(eq3)); memset(teq3,0,sizeof(teq3));
     memset(bbq,0,sizeof(bbq)); memset(tbbq,0,sizeof(tbbq));
-    memset(qt1,0,sizeof(qt1)); memset(tqt1,0,sizeof(tqt1));
-    memset(qt2,0,sizeof(qt2)); memset(tqt2,0,sizeof(tqt2));
-    memset(qt3,0,sizeof(qt3)); memset(tqt3,0,sizeof(tqt3));
-    memset(qt4,0,sizeof(qt4)); memset(tqt4,0,sizeof(tqt4));
+    //memset(qt1,0,sizeof(qt1)); memset(tqt1,0,sizeof(tqt1));
```
Comment (veprbl) [inline StRoot/StEvent/StTriggerData2022.cxx]: Shouldn't we zero those fields anyway? Ideally, we'd remove the fields that are not present.

File: StRoot/StEvent/StTriggerData2022.cxx [hunk 4]
```diff
@@ -192,15 +194,17 @@ void StTriggerData2022::readData(const TriggerDataBlk2022* data, int bs) {
...
         if (mEQ3[i]) decodeQT(mEQ3[i]->length/4, mEQ3[i]->data, eq3[i], teq3[i]);
         if (mBBQ[i]) decodeQT(mBBQ[i]->length/4, mBBQ[i]->data, bbq[i], tbbq[i]);
-        if (mQT1[i]) decodeQT(mQT1[i]->length/4, mQT1[i]->data, qt1[i], tqt1[i]);
-        if (mQT2[i]) decodeQT(mQT2[i]->length/4, mQT2[i]->data, qt2[i], tqt2[i]);
-        if (mQT3[i]) decodeQT(mQT3[i]->length/4, mQT3[i]->data, qt3[i], tqt3[i]);
-        if (mQT4[i]) decodeQT(mQT4[i]->length/4, mQT4[i]->data, qt4[i], tqt4[i]);
+	//if (mQT1[i]) decodeQT(mQT1[i]->length/4, mQT1[i]->data, qt1[i], tqt1[i]);
+        //if (mQT2[i]) decodeQT(mQT2[i]->length/4, mQT2[i]->data, qt2[i], tqt2[i]);
+        //if (mQT3[i]) decodeQT(mQT3[i]->length/4, mQT3[i]->data, qt3[i], tqt3[i]);
+        //if (mQT4[i]) decodeQT(mQT4[i]->length/4, mQT4[i]->data, qt4[i], tqt4[i]);
```
Comment (veprbl) [inline StRoot/StEvent/StTriggerData2022.cxx]:
```suggestion
	// Crates 1 through 4 were not used in Run22
```
Remove the commented code, replace with human-readable comment

Comment (veprbl) [reply StRoot/StEvent/StTriggerData2022.cxx]: Or "QT boards were not used", you would know best, of course

File: StRoot/StEvent/StTriggerData2022.cxx (hunk 3 above)
Comment (akioogawa) [reply StRoot/StEvent/StTriggerData2022.cxx]: qt1~4 arrays are commented out. So cannot be zeroed.

File: StRoot/StEvent/StTriggerData2022.cxx (hunk 1 above)
Comment (akioogawa) [reply StRoot/StEvent/StTriggerData2022.cxx:1409]: 0 for BBQ was my backdoor for accessing BBQ raw data. I can remove but rather keep It. As for 1~4, It should return 0, and will return 0.

File: StRoot/StEvent/StTriggerData2022.cxx (hunk 2 above)
Comment (akioogawa) [reply StRoot/StEvent/StTriggerData2022.cxx:1427]: same as above

**Review by klendathu2k — APPROVED** [review]
Concerns have been addressed.  Approved.

File: StRoot/StEvent/StTriggerData2022.cxx [hunk 5]
```diff
@@ -144,15 +144,17 @@ void StTriggerData2022::readData(const TriggerDataBlk2022* data, int bs) {
...
     memset(eq3,0,sizeof(eq3)); memset(teq3,0,sizeof(teq3));
     memset(bbq,0,sizeof(bbq)); memset(tbbq,0,sizeof(tbbq));
-    memset(qt1,0,sizeof(qt1)); memset(tqt1,0,sizeof(tqt1));
-    memset(qt2,0,sizeof(qt2)); memset(tqt2,0,sizeof(tqt2));
-    memset(qt3,0,sizeof(qt3)); memset(tqt3,0,sizeof(tqt3));
-    memset(qt4,0,sizeof(qt4)); memset(tqt4,0,sizeof(tqt4));
+    //memset(qt1,0,sizeof(qt1)); memset(tqt1,0,sizeof(tqt1));
+    //memset(qt2,0,sizeof(qt2)); memset(tqt2,0,sizeof(tqt2));
+    //memset(qt3,0,sizeof(qt3)); memset(tqt3,0,sizeof(tqt3));
+    //memset(qt4,0,sizeof(qt4)); memset(tqt4,0,sizeof(tqt4));
```
Comment (veprbl) [inline StRoot/StEvent/StTriggerData2022.cxx]:
```suggestion
```

File: StRoot/StEvent/StTriggerData2022.h [hunk 6]
```diff
@@ -269,11 +269,13 @@ class StTriggerData2022 : public StTriggerData {
...
     QTBlock2022*      mEQ1[11];	//!
     QTBlock2022*      mEQ2[11];	//!
-    unsigned short mxq[11][16][32],eq3[11][16][32],bbq[11][16][32],qt1[11][16][32],qt2[11][16][32],qt3[11][16][32],qt4[11][16][32],
-    eq1[11][16][32],eq2[11][16][32]; //!
-    unsigned char tmxq[11][16][32],teq3[11][16][32],tbbq[11][16][32],tqt1[11][16][32],tqt2[11][16][32],tqt3[11][16][32],tqt4[11][16][32],
-    teq1[11][16][32],teq2[11][16][32]; //!
-    
+    QTBlock2022*      mEQ4[11];	//!
+    unsigned short mxq[11][16][32],eq3[11][16][32],bbq[11][16][32],
+      //qt1[11][16][32],qt2[11][16][32],qt3[11][16][32],qt4[11][16][32],
```
Comment (veprbl) [inline StRoot/StEvent/StTriggerData2022.h]:
```suggestion
```

File: StRoot/StEvent/StTriggerData2022.h [hunk 7]
```diff
@@ -269,11 +269,13 @@ class StTriggerData2022 : public StTriggerData {
...
+      eq1[11][16][32],eq2[11][16][32],eq4[11][16][32]; //!
+    unsigned char tmxq[11][16][32],teq3[11][16][32],tbbq[11][16][32],
+      //tqt1[11][16][32],tqt2[11][16][32],tqt3[11][16][32],tqt4[11][16][32],
```
Comment (veprbl) [inline StRoot/StEvent/StTriggerData2022.h]:
```suggestion
```

File: StRoot/StEvent/StTriggerData2022.cxx [hunk 8]
```diff
@@ -1421,16 +1423,17 @@ unsigned short StTriggerData2022::fmsADC(int crt, int adr, int ch, int prepost)
...
         switch(crt){
             case 0: return tbbq[buffer][adr][ch];
-            case 1: return tqt1[buffer][adr][ch];
-            case 2: return tqt2[buffer][adr][ch];
-            case 3: return tqt3[buffer][adr][ch];
-            case 4: return tqt4[buffer][adr][ch];
+	      //case 1: return tqt1[buffer][adr][ch];
+	      //case 2: return tqt2[buffer][adr][ch];
+	      //case 3: return tqt3[buffer][adr][ch];
+	      //case 4: return tqt4[buffer][adr][ch];
```
Comment (veprbl) [inline StRoot/StEvent/StTriggerData2022.cxx]:
```suggestion
```

File: StRoot/StEvent/StTriggerData2022.cxx [hunk 9]
```diff
@@ -1403,16 +1404,17 @@ unsigned short StTriggerData2022::mtd4AtAddress(int address, int prepost) const
...
         switch(crt){
 	    case 0: return bbq[buffer][adr][ch];
-	    case 1: return qt1[buffer][adr][ch];
-            case 2: return qt2[buffer][adr][ch];
-            case 3: return qt3[buffer][adr][ch];
-            case 4: return qt4[buffer][adr][ch];
+	    //case 1: return qt1[buffer][adr][ch];
+	    //case 2: return qt2[buffer][adr][ch];
+	    //case 3: return qt3[buffer][adr][ch];
+	    //case 4: return qt4[buffer][adr][ch];
```
Comment (veprbl) [inline StRoot/StEvent/StTriggerData2022.cxx]:
```suggestion
```

File: StRoot/StEvent/StTriggerData2022.cxx [hunk 10]
```diff
@@ -192,15 +194,14 @@ void StTriggerData2022::readData(const TriggerDataBlk2022* data, int bs) {
...
         j=offlen[y22EQ2_CONF_NUM].length; if (j>12){mEQ2[i] = (QTBlock2022*     )((char*)mData + offlen[y22EQ2_CONF_NUM].offset); swapRawDet((DataBlock2022*)mEQ2[i],y22EQ2_CONF_NUM,j,bs);}
+        j=offlen[y22EQ4_CONF_NUM].length; if (j>12){mEQ4[i] = (QTBlock2022*     )((char*)mData + offlen[y22EQ4_CONF_NUM].offset); swapRawDet((DataBlock2022*)mEQ4[i],y22EQ4_CONF_NUM,j,bs);}
         if (mMXQ[i]) decodeQT(mMXQ[i]->length/4, mMXQ[i]->data, mxq[i], tmxq[i]);
         if (mEQ3[i]) decodeQT(mEQ3[i]->length/4, mEQ3[i]->data, eq3[i], teq3[i]);
         if (mBBQ[i]) decodeQT(mBBQ[i]->length/4, mBBQ[i]->data, bbq[i], tbbq[i]);
-        if (mQT1[i]) decodeQT(mQT1[i]->length/4, mQT1[i]->data, qt1[i], tqt1[i]);
-        if (mQT2[i]) decodeQT(mQT2[i]->length/4, mQT2[i]->data, qt2[i], tqt2[i]);
-        if (mQT3[i]) decodeQT(mQT3[i]->length/4, mQT3[i]->data, qt3[i], tqt3[i]);
-        if (mQT4[i]) decodeQT(mQT4[i]->length/4, mQT4[i]->data, qt4[i], tqt4[i]);
+	// Crates 1 through 4 were not used in Run22
```
Comment (veprbl) [inline StRoot/StEvent/StTriggerData2022.cxx]:
```suggestion
	// QT boards were not used in Run22
```

Comment (veprbl) [reply StRoot/StEvent/StTriggerData2022.cxx]: Or
```suggestion
```

**Comment by plexoos** [comment]
I highly recommend to accept the changes by Dmitry K. clean and readable code is something we care about.
If no other comments we can merge ASAP.

**Review by plexoos — APPROVED** [review]

**Comment by plexoos** [comment]
OMG see https://github.com/star-bnl/star-sw/pull/447#issuecomment-1322323042

Screen Shot 2022-12-01 at 9 53 56 AM

There is a good reason why they have the "Add suggestion to batch" button

**Comment by veprbl** [comment]
> OMG see [#447 (comment)](https://github.com/star-bnl/star-sw/pull/447#issuecomment-1322323042)
> 
> Screen Shot 2022-12-01 at 9 53 56 AM
> 
> There is a good reason why they have the "Add suggestion to batch" button

Let's address this in https://github.com/star-bnl/star-sw/pull/453

**Pull request merged**